### PR TITLE
Buffer copy improvements

### PR DESCRIPTION
**Pull request by zvookin**
Added Metal support.

Fixed a bug in OpenCL implementation.

Make case for CUDA and OpenCL do device to device instead of host to device.

### CODE REVIEW

File: src/runtime/metal.cpp [hunk 1]
```diff
@@ -816,6 +848,98 @@ WEAK int halide_metal_device_and_host_free(void *user_context, struct halide_buf
...
+    // We only handle copies to cuda or to host
+    halide_assert(user_context, dst_device_interface == NULL ||
+                  dst_device_interface == &metal_device_interface);
+
+    if (src->device_dirty() &&
+        src->device_interface != &metal_device_interface) {
+        halide_assert(user_context, dst_device_interface == &metal_device_interface);
+        // If the source is not metal or host memory, ask the source
+        // device interface to copy to dst host memory first.
+        // TODO: this probably does not set the appropriate dirty bit. Probably needs to call up to generic routine. Fix CUDA too.
```
Comment (shoaibkamil) [inline src/runtime/metal.cpp]: I agree that this should set the dirty bit.

**Comment by steven-johnson** [comment]
Failure is unrelated and should not block landing.

**Comment by zvookin** [comment]
I'm doing a modest amount of work on halide_buffer_copy to make it do device-to-device copies more often and to not recurse into itself. Hopefully have a revised version of this in the next few days.

File: src/runtime/device_interface.cpp [hunk 2]
```diff
@@ -20,10 +20,10 @@ struct device_handle_wrapper {
...
 // TODO: Coarser grained locking, also consider all things that need
 // to be atomic with respect to each other. At present only
-// halide_copy_to_host and halide_copy_to_device are atomic with
-// respect to each other. halide_device_malloc and halide_device_free
-// are also candidates, but to do so they likely need to be able to do
-// a copy internaly as well.
+// halide_copy_to_host, halide_copy_to_device, and halide_buffer_copy
+// are atomic with respect to each other. halide_device_malloc and
+// halide_device_free are also candidates, but to do so they likely
+// need to be able to do a copy internaly as well.
```
Comment (abadams) [inline src/runtime/device_interface.cpp]: Was pre-existing, but internaly -> internally

File: src/runtime/device_interface.cpp [hunk 3]
```diff
@@ -440,66 +440,24 @@ WEAK void halide_device_host_nop_free(void *user_context, void *obj) {
...
 
 WEAK int halide_default_buffer_copy(void *user_context, struct halide_buffer_t *src,
-                                    const struct halide_device_interface_t *dst_device_interface,
-                                    struct halide_buffer_t *dst) {
-
-    const bool from_device = src->device && (src->device_dirty() || !src->host);
-    const bool to_device = dst_device_interface;
-    const bool from_host = !from_device;
-    const bool to_host = !to_device;
+                                           const struct halide_device_interface_t *dst_device_interface,
```
Comment (abadams) [inline src/runtime/device_interface.cpp]: wonky indentation?

File: src/runtime/device_interface.cpp [hunk 4]
```diff
@@ -509,54 +467,131 @@ WEAK int halide_buffer_copy(void *user_context, struct halide_buffer_t *src,
...
+            debug(user_context) << "halide_buffer_copy_already_locked: failing due to need for temp buffer.\n";
+            return halide_error_code_incompatible_device_interface;
+        }
+
+        if (to_host) {
+            debug(user_context) << "halide_buffer_copy_already_locked: to host case.\n";
+            err = src->device_interface->impl->buffer_copy(user_context, src, NULL, dst);
+            // Return on success or an error indicating something other
+            // than not handling this case went wrong.
+            if (err == halide_error_code_incompatible_device_interface) {
```
Comment (abadams) [inline src/runtime/device_interface.cpp]: Should these be if (err != 0) ? What if the implementation throws some other error.

Comment (abadams) [reply src/runtime/device_interface.cpp]: Wait, nvm, this is handling a specific error.

File: src/runtime/opencl.cpp [hunk 5]
```diff
@@ -806,11 +806,11 @@ WEAK int do_multidimensional_copy(void *user_context, ClContext &ctx,
             err = clEnqueueCopyBuffer(ctx.cmd_queue, ((device_handle *)c.src)->mem, ((device_handle *)c.dst)->mem,
                                       src_idx + ((device_handle *)c.src)->offset, dst_idx  + ((device_handle *)c.dst)->offset,
                                       c.chunk_size, 0, NULL, NULL);
-        } else if (c.dst != c.src) {
+        } else if ((c.dst + dst_idx) != (c.src + src_idx)) {
```
Comment (abadams) [inline src/runtime/opencl.cpp]: oops

File: src/runtime/cuda.cpp [hunk 6]
```diff
@@ -727,18 +731,16 @@ WEAK int halide_cuda_buffer_copy(void *user_context, struct halide_buffer_t *src
...
-        int err = src->device_interface->impl->buffer_copy(user_context, src, NULL, dst);
-        if (err) return err;
-        // Now just copy from src to host
-        src = dst;
+        // This is handled at the higher level.
+        return halide_error_code_incompatible_device_interface;
     }
 
-    bool from_host = !src->device_dirty() && src->host;
+    bool from_host = (src->device_interface != &cuda_device_interface) ||
```
[the remark this answers is not in the document]
Comment (zvookin) [reply src/runtime/cuda.cpp]: I think the place to go is to push the logic up into the device_interface.cpp layer and just have the device specific part support three variants of memcpy (device - >device, device -> host-pointer, host-pointer -> device). For now I'm just going to leave it.

**Comment by zvookin** [comment]
Test failures are unrelated. One is simd_op_check, the other is failure to find packages.

**Pull request merged**